### PR TITLE
Add DDAssert* methods

### PR DESCRIPTION
**Pull request by maxep**
### What and why?

Custom assertions does not follow a strict pattern, making tests implementation somehow tricky. Some were functions, but other methods, some were throwing but other not.
In the effort of sanitising and sharing test utilities with modules, we should provide a single pattern for doing assertions.

### How?

`DDAssert*` functions can be used in addition to `XCTAssert*` to perform specific assertions, it follows the exact same pattern for a seamless devx. Here the list of assertions included in this PR:
- `DDAssertReflect`: in replacement of `EquatableInTests`
- `DDAssertNotReflect`
- `DDAssertJSONEqual`: in replacement of `AssertEncodedRepresentationsEqual`
- `DDAssertJSONNotEqual`
- `DDAssertDictionariesEqual`: in replacement of `AssertDictionariesEqual`
- `DDAssertDictionariesNotEqual`

**Why replacing `EquatableInTests`?** The `EquatableInTests` was not a real equality but used reflection and `CustomStringConvertible` to compare, this specificity could be confusing. The `DDAssertReflect` now explicitly described the assertion made.

### Review checklist
- [x] Feature or bugfix MUST have appropriate tests (unit, integration)
- [ ] Make sure each commit and the PR mention the Issue number or JIRA reference
- [ ] Add CHANGELOG entry for user facing changes

### Custom CI job configuration (optional)
- [x] Run unit tests
- [x] Run integration tests
- [x] Run smoke tests

### CODE REVIEW

File: TestUtilities/Sources/TestUtilities/Helpers/DDAssert.swift [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+ */
+
+import Foundation
+import XCTest
+
+private enum DDAssertError: Error {
+    case expectedFailure(String, keyPath: [String] = [])
+}
+
+/// This function emits a test failure if the general expression throw any error.
```
Comment (ncreated) [inline TestUtilities/Sources/TestUtilities/Helpers/DDAssert.swift:17]:
```suggestion
/// This function emits a test failure if the general expression throws any error.
```

File: TestUtilities/Sources/TestUtilities/Helpers/DDAssert.swift [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+            try _DDAssertReflect(expression1(), expression2())
+        } catch DDAssertError.expectedFailure {
+            return // expected
+        }
+
+        throw DDAssertError.expectedFailure("Reflections are equal")
+    }
+}
+
+private func _DDAssertJSONEqual<T, U>(_ expression1: @autoclosure () throws -> T, _ expression2: @autoclosure () throws -> U, _ message: @autoclosure () -> String = "", file: StaticString = #filePath, line: UInt = #line) throws where T: Encodable, U: Encodable {
```
Comment (ncreated) [inline TestUtilities/Sources/TestUtilities/Helpers/DDAssert.swift:127]: These parameters aren't necessary as we evaluate failures in `_DDEvaluateAssertion` 🙌 
```suggestion
private func _DDAssertJSONEqual<T, U>(_ expression1: @autoclosure () throws -> T, _ expression2: @autoclosure () throws -> U) throws where T: Encodable, U: Encodable {
```

File: TestUtilities/Sources/TestUtilities/Helpers/DDAssert.swift [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+        break // other than dictionary or set, continue...
+    }
+
+    for (child1, child2) in zip(mirror1.children, mirror2.children) { // compare each child
+        let key = child1.label.map { [$0] } ?? []
+        try _DDAssertReflect(child1.value, child2.value, keyPath: keyPath + key)
+    }
+}
+
+public func DDAssertReflect<T>(_ expression1: @autoclosure () throws -> T, _ expression2: @autoclosure () throws -> T, _ message: @autoclosure () -> String = "", file: StaticString = #filePath, line: UInt = #line) {
```
Comment (ncreated) [inline TestUtilities/Sources/TestUtilities/Helpers/DDAssert.swift:109]: nit/ The name could be quite vague, hence adding a comment could be priceless for newcomers.

```suggestion
/// Asserts that both expressions equal by using reflection and comparing their mirrors.
public func DDAssertReflect<T>(_ expression1: @autoclosure () throws -> T, _ expression2: @autoclosure () throws -> T, _ message: @autoclosure () -> String = "", file: StaticString = #filePath, line: UInt = #line) {
```

File: TestUtilities/Sources/TestUtilities/Helpers/DDAssert.swift [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+    _DDEvaluateAssertion(message: message(), file: file, line: line) {
+        try _DDAssertReflect(expression1(), expression2())
+    }
+}
+
+public func DDAssertNotReflect<T>(_ expression1: @autoclosure () throws -> T, _ expression2: @autoclosure () throws -> T, _ message: @autoclosure () -> String = "", file: StaticString = #filePath, line: UInt = #line) {
```
Comment (ncreated) [inline TestUtilities/Sources/TestUtilities/Helpers/DDAssert.swift:115]: nit/
```suggestion
/// Asserts that both expressions do not equal by using reflection and comparing their mirrors.
public func DDAssertNotReflect<T>(_ expression1: @autoclosure () throws -> T, _ expression2: @autoclosure () throws -> T, _ message: @autoclosure () -> String = "", file: StaticString = #filePath, line: UInt = #line) {
```

**Review by ncreated — APPROVED** [review]
This is great 🚀. I agree, the way we were serving this utilities evolved in a very inconsequent way over years. Now it's unified, easy to discover and obvious to extend 👍🏅.

nit/ I think this bullet point in the PR description is incorrect:
> `DDAssertNotReflect` in replacement of `AssertEncodedRepresentationsEqual`

**Comment by maxep** [comment]
image

**Pull request merged**